### DISCUSSION TITLE
[Enhancement] More informative signaling error messages

### DISCUSSION THREAD

**Discussion by suggestedfixes** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-js#216] [Ideas]
From time to time, we receive an error event that was implemented in the js sdk.
For example, we get status code of 400 from time to time.
From this document
https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-websocket-apis-1.html, we know that it is either ClientLimitExceededException or InvalidArgumentException. But we can only guess what the error is. It would be nice to reinterpret the Websocket Error into an Error with meaningful name.

**Comment by lherman-cs** [comment]
@suggestedfixes Do you mind to describe what you mean by "guess" in the case? As you've mentioned here that a 400 error can be either ClientLimitExceededException or InvalidArgumentException. Doesn't that give you the error kinds?

**Comment by MushMal** [comment]
@suggestedfixes also, which API are you referring to? The data plane APIs are limited.

**Comment by suggestedfixes** [comment]
@lherman-cs @MushMal 
As described in issue #41 
The Error Format caught is
```
Event {isTrusted: true, type: "error", target: WebSocket, currentTarget: WebSocket, eventPhase: 2, …}
bubbles: false
cancelBubble: false
cancelable: false
composed: false
currentTarget: WebSocket {url: "wss://v-dc297268.kinesisvideo.us-east-2.amazonaws.…36a2fafd25a96fa0aeac3dd0&X-Amz-SignedHeaders=host", readyState: 3, bufferedAmount: 0, onopen: null, onerror: null, …}
defaultPrevented: false
eventPhase: 0
isTrusted: true
path: []
returnValue: true
srcElement: WebSocket {url: "wss://v-dc297268.kinesisvideo.us-east-2.amazonaws.…36a2fafd25a96fa0aeac3dd0&X-Amz-SignedHeaders=host", readyState: 3, bufferedAmount: 0, onopen: null, onerror: null, …}
target: WebSocket {url: "wss://v-dc297268.kinesisvideo.us-east-2.amazonaws.…36a2fafd25a96fa0aeac3dd0&X-Amz-SignedHeaders=host", readyState: 3, bufferedAmount: 0, onopen: null, onerror: null, …}
timeStamp: 9966.775000095367
type: "error"
__proto__: Event
```

Now besides the client limit exceeded for 400, another thing that could go wrong is the signing of the endpoint failed due to invalid arguments.

I couldn't find that data structure that holds the Amazon error type. From my limited javascript knowledge, maybe extending the Websocket object to include more attributes such as aws-error-type might work?

Also from https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-websocket-apis-1.html. I might have interpreted incorrectly, seems like there is a method to capture the Exception type. I'm likely not aware of the technique that can capture the Exception Name and string.

**Comment by suggestedfixes** [comment]
@MushMal I'm referring to SignalingClient.on('error', handler); API

**Comment by hassanctech** [comment]
This will require some deeper investigation on our part, we will get back to you here once we've had a chance to investigate this.
